src/pointer/ptr.rs

```diff
@@ -0,0 +1,1010 @@
+        pub(crate) const unsafe fn assume_valid(
+            self,
+        ) -> Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, { invariant::Valid }> {
+            // SAFETY: The caller promises that `self`'s referent is bit-valid
+            // for `T`.
+            unsafe { Ptr::from_ptr(self) }
+        }
+
+        /// Forgets that `Ptr`'s referent is validly-aligned for `T`.
+        #[doc(hidden)]
+        #[inline]
+        pub const fn forget_aligned(
+            self,
+        ) -> Ptr<'a, T, ALIASING_INVARIANT, { invariant::AnyAlignment }, VALIDITY_INVARIANT>
+        {
+            // SAFETY: `AnyAlignment` is less restrictive than `Aligned`.
+            unsafe { Ptr::from_ptr(self) }
+        }
+
+        /// Forgets that `Ptr`'s referent is bit-valid for `T`.
+        #[doc(hidden)]
+        #[inline]
+        pub const fn forget_valid(
+            self,
+        ) -> Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, { invariant::AsInitialized }>
+        {
+            // SAFETY: `AnyValidity` is less restrictive than `Valid`.
+            unsafe { Ptr::from_ptr(self) }
+        }
+    }
+}
+
+/// Casts of the referent type.
+mod _casts {
+    use super::*;
+
+    impl<
+            'a,
+            T,
+            const ALIASING_INVARIANT: invariant::Aliasing,
+            const ALIGNMENT_INVARIANT: invariant::Alignment,
+            const VALIDITY_INVARIANT: invariant::Validity,
+        > Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, VALIDITY_INVARIANT>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Casts to a different (unsized) target type.
+        ///
+        /// # Safety
+        ///
+        /// The caller promises that
+        /// - `cast(p)` is implemented exactly as follows: `|p: *mut T| p as
+        ///   *mut U`.
+        /// - The size of the object referenced by the resulting pointer is less
+        ///   than or equal to the size of the object referenced by `self`.
+        #[doc(hidden)]
+        #[inline]
+        pub unsafe fn cast_unsized<U: 'a + ?Sized, F: FnOnce(*mut T) -> *mut U>(
+            self,
+            cast: F,
+        ) -> Ptr<'a, U, ALIASING_INVARIANT, { invariant::AnyAlignment }, { invariant::AnyValidity }>
+        where
+            U: 'a,
+        {
+            let ptr = cast(self.as_non_null().as_ptr());
+
+            // SAFETY: Caller promises that `cast` is just an `as` cast. We call
+            // `cast` on `self.ptr.as_ptr()`, which is non-null by construction.
+            let ptr = unsafe { NonNull::new_unchecked(ptr) };
+
+            // SAFETY:
+            // 0. By invariant on `self` and contract on `cast`, `ptr` is
+            //    derived from some valid Rust allocation, `A`.
+            // 1. By invariant on `self` and contract on `cast`, `ptr` has valid
+            //    provenance for `A`.
+            // 2. By invariant on `self` and contract on `cast`, `ptr` addresses
+            //    a byte range which is entirely contained in `A`.
```

Comment:
Maybe call out explicitly that this is a result of the fact that `cast` cannot produce a pointer which addresses a larger-sized object, and so the addressed bytes are a subset of the bytes addressed by `self`, and thus are also contained in `A`? Same with points 3 and 4.